libsupport/include/katana/ProgressTracer.h

```diff
@@ -0,0 +1,222 @@
+
+class ProgressScope;
+class ProgressSpan;
+class ProgressContext;
+
+class KATANA_EXPORT ProgressTracer {
+  virtual std::shared_ptr<ProgressSpan> StartSpan(
+      const std::string& span_name, std::shared_ptr<ProgressSpan> child_of) = 0;
+
+  ProgressScope SetActiveSpan(std::shared_ptr<ProgressSpan> span);
+
+protected:
+  ProgressTracer(uint32_t host_id, uint32_t num_hosts)
+      : host_id_(host_id), num_hosts_(num_hosts) {}
+
+public:
+  virtual ~ProgressTracer() = default;
+
+  static ProgressTracer& GetProgressTracer() { return *tracer_; }
+  static void SetProgressTracer(std::unique_ptr<ProgressTracer> tracer);
+
+  // StartActiveSpan creates a new span. If there is not an active span,
+  // create a new top-level span. Otherwise, this function creates a child
+  // span of the active span. The returned scope will finish
+  // the span on close
+  ProgressScope StartActiveSpan(const std::string& span_name);
+  ProgressScope StartActiveSpan(
+      const std::string& span_name, const ProgressContext& child_of);
+
+  /// Create a new top level span if ignore_active_span=true and
+  /// no child_of value is given
+  /// Otherwise creates a child span of the child_of span or active span
+  /// Should only be used to handle multiple active spans simultaneously
+  virtual std::shared_ptr<ProgressSpan> StartSpan(
+      const std::string& span_name, const ProgressContext& child_of) = 0;
+
+  /// Calls finish on the active span, its parent is set to be active
+  /// Primarily for internal use only, preferred to use ProgressSpan's Finish function
+  void FinishActiveSpan();
+
+  // For passing spans across process/host boundaries
+  // These functions are needed, but are implemented now for
+  // debugging purposes, they will be replaced
+  // Extract returns a nullptr on failure
+  virtual std::string Inject(const ProgressContext& ctx) = 0;
+  virtual std::unique_ptr<ProgressContext> Extract(
+      const std::string& carrier) = 0;
+
+  /// Get the current scope’s span to add tagging/logging without having
+  /// to pass around scopes/spans as parameters
+  /// If there is no active span, an unnamed root span of a new trace
+  /// is created (in this case the program is probably not using tracing)
+  virtual ProgressSpan& GetActiveSpan();
+  bool HasActiveSpan() { return active_span_ != nullptr; }
+  uint32_t GetHostID() { return host_id_; }
+  uint32_t GetNumHosts() { return num_hosts_; }
+
+  /// Close is called when a tracer is finished processing spans
+  /// This should be called to ensure any and all buffered spans are
+  /// flushed
+  virtual void Close() = 0;
+
+private:
+  static std::unique_ptr<ProgressTracer> tracer_;
+
+  std::shared_ptr<ProgressSpan> active_span_ = nullptr;
+  uint32_t host_id_;
+  uint32_t num_hosts_;
+  std::shared_ptr<ProgressSpan> default_active_span_ = nullptr;
+};
+
+class KATANA_EXPORT ProgressScope {
+  friend class ProgressTracer;
+
+  ProgressScope(std::shared_ptr<ProgressSpan> span) : span_(span) {}
```

Comment:
```ProgressScope(std::shared_ptr<ProgressSpan> span): span_(std::move(span)) {}```

There are a few places in this PR where we can avoid extra copies of shared_ptrs